### ISSUE TITLE
Proposal: Front End Engineering Week

### ISSUE THREAD

**Issue by ronanyeah** [foundersandcoders/master-reference#264]
If there was a week that focused on front end engineering, what would that look like? I think it is important that this would remain front end framework agnostic. Partly because things move quickly in the JS (and compile-to-JS) world and partly because [holy](https://facebook.github.io/react/) [moly](https://angular.io/) [there](http://elm-lang.org/) [are](https://vuejs.org/) [a](http://riotjs.com/) [lot](https://www.emberjs.com/) [of](http://www.purescript.org/) [options](https://clojurescript.org/).

(I realise that the first 8 weeks is totally packed. This is hypothetical.)

Topics:
- [Webpack](https://webpack.js.org/)
- [Babel](https://babeljs.io)
- [ES6](http://es6-features.org)
- [Promises](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Promise)
- [Async/Await](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function)
- [import](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Statements/import)/[export](https://developer.mozilla.org/en/docs/web/javascript/reference/statements/export)

Exercises:
- 'Pick a Framework' Day: Teams pick a framework/library each and present a 'Hello World' in it.

Learning Outcomes:
- How and WHY to bundle your code.
- How and WHY to transpile your code.
- Comfort with ES6.

**Comment by jsms90** [comment]
Thanks for getting so engaged with this @ronanyeah! It's great to hear opinions from our wider community :tada: 

Can I ask whether this is intended as a counter proposal to the one about promises #262 or the HTML/CSS one #258? I wonder whether it might be more beneficial to communicate with @oliverjam on his proposals? Otherwise, I'm afraid that this one just competes unnecessarily with them?

As for the content:
+ Theoretically, the "topics" bit could go somehwere in the first half of the course.
+ But if the "exericses" bit belongs anywhere, "Pick a Framework' Day" should go in the second half of the course. I would never encourage people to spend the first half of the course learning different things from each other. FAC is about building communities & you can't do that when everyone is learning separate things as individuals.

**Comment by ronanyeah** [comment]
This issue is for discussing a theoretical week on front end engineering in JavaScript. HTML and CSS aren't really related. Promises could possibly be drawn into this discussion as the main reason they aren't taught are for browser support concerns. 

This could absolutely be post week 8 content. But again, I just want a discussion.

**Comment by ronanyeah** [comment]
I don't agree that you can't have a community where people have different interests.

**Comment by jsms90** [comment]
There's nothing wrong with having different _interests_. But the "jigsaw clasroom" in research afternoon consistently presents problems. Different people end up being aware of or missing out on different things.

**Comment by ronanyeah** [comment]
@jsms90 Oh interesting. Yeah I was only suggesting that as the jigsawing seemed to be a common exercise. I'll strike it out.

**Comment by jsms90** [comment]
IMO we have 2 proposals for introducing new technologies when we don't even teach people the basics of frontend on this course https://github.com/foundersandcoders/master-reference/issues/259#issuecomment-302859046. And I believe it's a _huge_ problem that we relegate that learning to the pre-requisites!!

We're finally having a serious discussion about putting in a front end week :heart_eyes: :heart_eyes: :heart_eyes: but I honestly don't believe that the level people come to us with makes Webpack, Babel, etc the appropriate place to start. I would very much consider these ["Advanced" concepts](https://github.com/foundersandcoders/master-reference/blob/master/coursebook/fragments/product-build/tech-choices.md#advanced-javascript) by comparison to all the things that our students are lacking on the course.

I totally agree with ES6 & async/await. I would like us to do promises. But I don't think we should be doing it until we've figured out how to teach callbacks in the first place (but that's part of another proposal anyway).

There are so many topics, big and small, I don't even know where to start...
+ Traversing the DOM
+ Scope & the `this` keyword
+ Passing by reference vs passing by value
+ Understanding closures
+ Hoisting
+ ES6 :+1: 
+ Client-side storage
+ Client-side form validation
+ Caching
+ Using dev tools effectively e.g. for debugging

Also, how about:
+ Choosing libraries to use
+ Creating your own library

**Comment by eliascodes** [comment]
@jsms90 

> This could absolutely be post week 8 content.

I don't see an issue with this being in the second half. The stuff you're talking about should be in the first half. We'd not be starting with webpack.

**Comment by jsms90** [comment]
Good point @eliascodes 

> This could absolutely be post week 8 content. But again, **I just want a discussion.**

So @ronanyeah, I suggest that you remove the `proposal` label, and put `discussion`? We already have a lot to get through tomorrow and I'm not sure whether discussions will make it into the time that we have. Either way, I'm sure it would help @bradreeder to have a clearer idea of what's on the table.

Or you could keep this as a proposal, but edit it so that it reflects the intended "What" and "How".

For example, it could be:
**What**
<Suggest where this should go within weeks 1-8>

**How**
+ a set of students from the current cohort(s) assign themselves to this week and follow the [current procedure](https://github.com/foundersandcoders/master-reference/blob/master/curriculum-planning/planning-weeks.md) for designing a week, like everyone else
+ X alumni volunteer to assist and/or review the students' plans before they present their week outline
+ X alumni volunteer to assist and/or review the students progress with the creation of the workshops themselves

It might be difficult to vote on a proposal, where people don't know what the suggestion is for the content of the week. Or it might be that people trust the students/alumni to make something, and believe that something is better than the nothing that we have now.

I guess it's your choice, @ronanyeah

**Comment by m4v15** [comment]
I really like the idea of some sort of webpack/bundling/babel workshop - literally everything uses it, and it wouldn't really take that long to get everyone on the same page with it.

I agree that it's not necessary early on, and might actually confuse things at that stage, but I think by the second half of the course, a quick workshop on how to structure and engineer frontend would be **very** beneficial and relatively easy to do with everyone comfortable with node/npm.

**Comment by jsms90** [comment]
@m4v15 Have you seen @eliascodes' updates to the [tech choices doc](https://github.com/foundersandcoders/master-reference/blob/master/coursebook/fragments/product-build/tech-choices.md). It a really nice start on guidance into choosing new tech :heart:

**Comment by ronanyeah** [comment]
@jsms90 That document says that webpack is 'Ubiquitous in industry'. Bundling and minification are not 'tech choices', they are essential front end development skills, and I think they have a place in the first 8 weeks. The entire workshop content can be summarised in 19 lines, it could basically be a morning challenge.

```
// webpack.config.js
const { resolve } = require('path');

const outputPath = resolve(__dirname);

module.exports = {
  entry: './index.js',
  output: {
    path: outputPath,
    filename: 'bundle.js'
  }
};
```

```
// foo.js
export default 'yeah';

const otherThing = 'other thing';

export { otherThing };
```

```
// index.js
import yeah, { otherThing } from './foo.js';

console.log(yeah, otherThing); // 'yeah other thing'
```

**Comment by ronanyeah** [comment]
@jsms90 And yeah I think this is too broad for tomorrow, I'll open another issue to propose webpack/bundling and table this one.

**Comment by jsms90** [comment]
True, true. Completely the wrong wording. React is a "tech choice", SASS is a choice, hapi and express are choices. Webpack is more of a tool in your toolbelt, right? Something which is becoming an increasingly necessary tool.

I would still consider it "advanced" in the context of the other things we do not even teach to students about the language they are using. Webpack probably has a place in the first 8 weeks! But I don't think it should be in the next iteration. I think we need to put other, even more fundamental things into the first 8 weeks considering there is virtually no frontend :smile:

**Comment by ronanyeah** [comment]
@jsms90 I think 2 months is too long to be developing with either A. One massive script file. B. Multiple/many hard to follow script files.

**Comment by jsms90** [comment]
No matter how small webpack is, a morning challenge on that is a morning challenge slot that we don't have on something else

...although....morning challenge and then you get to use webpack....:thinking: NO JEN NO, put the shiny thing down

Seriously though, think about it like this. You have one week to teach students every single thing about frontend javascript (besides querying an API). Is webpack the one?

**Comment by ronanyeah** [comment]
@jsms90 One week? I don't think it should be taught in the first 4 weeks. You need npm scripts for it for one thing. It could be somewhere in the last 2 weeks since Hapi is likely getting chopped. It shouldn't even take a day. I'll be making the resources anyway probably.

**Comment by jsms90** [comment]
Sorry, you threw me off with the title of your proposal :stuck_out_tongue_winking_eye: 
> Front End Engineering Week

**Comment by ronanyeah** [comment]
@jsms90 Ha ok, right we can forget about this issue for now. I'm adding a bundling workshop proposal. :+1:

**Comment by jsms90** [comment]
@ronanyeah How about we change this back to a proposal and change the contents? If we're not yet ready i.e. have all the necessary resources together to deliver a solid frontend engineering week, how about we set a plan in motion for building one _now_ so that it _is_ ready by FAC12, FACN3 & FACG3?

### WHAT

Begin creating a week that focuses on front end engineering, which is front end framework agnostic. Partly because things move quickly in the JS (and compile-to-JS) world and partly because holy moly there are a lot of options.

There are however, things that are common across frameworks, such as bundling, minification, file versioning and progressive web app technologies.

HOW

Form a task group how commit to designing this week together.
They should follow the week planning process that aleady exists, by
+ first having a meeting together that follows [this outline](https://github.com/foundersandcoders/master-reference/blob/master/curriculum-planning/planning-weeks.md)
+ putting forward their proposal for the learning outcomes, project description & week's schedule
+ allowing a 1 month period for this proposal to be reviewed and commented on by the wider community
+ bringing the revised version to a vote
  + if passed, submit a PR to put this new week into `master-reference` (fragments folder? not into anyone's actual curriculum yet)
+ now having the approval of the FAC community, divide up the tasks and begin creating the morning challenges & workshops for the week
+ submit a proposal for discussion with FAC12, FACN3 & FACG3 for inclusion of this week in their curriculum(s)

**Comment by ronanyeah** [comment]
@jsms90 A bit late for adding more proposals. :smile: We can revisit this over the summer perhaps.

**Comment by jsms90** [comment]
@ronanyeah Just spoke to @bradreeder. I think it's still going on the agenda, even though it's a discussion.

We'll just have to see
